Hints.py

```diff
@@ -341,15 +348,44 @@ def get_woth_hint(spoiler, world, checked):
 
 
 def get_barren_hint(spoiler, world, checked):
+    if not hasattr(world, 'get_barren_hint_prev'):
+        world.get_barren_hint_prev = RegionRestriction.NONE
+
     areas = list(filter(lambda area:
-        area not in checked and \
-        not (world.barren_dungeon >= world.hint_dist_user['dungeons_barren_limit'] and \
-        world.empty_areas[area]['dungeon']),
+        area not in checked
+        and not (world.barren_dungeon >= world.hint_dist_user['dungeons_barren_limit'] and world.empty_areas[area]['dungeon']),
         world.empty_areas.keys()))
 
     if not areas:
         return None
 
+    # Randomly choose between overworld or dungeon
+    dungeon_areas = list(filter(lambda area: world.empty_areas[area]['dungeon'], world.empty_areas.keys()))
+    overworld_areas = list(filter(lambda area: not world.empty_areas[area]['dungeon'], world.empty_areas.keys()))
```

Comment:
```suggestion
    dungeon_areas = list(filter(lambda area: world.empty_areas[area]['dungeon'], areas))
    overworld_areas = list(filter(lambda area: not world.empty_areas[area]['dungeon'], areas))
```

At this point, `areas` is already defined and you should pull from that instead of `world.empty_areas.keys()`, otherwise you are throwing away the `checked` filtering as well as the barren dungeon limit.